File: recursion/quicksort.js
// quicksort :: Ord a => [a] -> [a]
const quicksort = xs => {
  if(!xs.length) {
    return [];
  }
  const [y, ...ys] = xs;
  const left = quicksort(ys.filter(x => x <= y));
  const right = quicksort(ys.filter(x => x > y));
  return [...left, y, ...right];
};

const n = 4000, test = Array(n).fill().map((_, i) => n - i);
const result = quicksort(test).filter((_, i) => i < 10);
console.log(result);

File: recursion/quicksort.hs
quicksort :: (Ord a) => [a] -> [a]
quicksort [] = []
quicksort (x:xs) =
  -- let left = quicksort [a | a <- xs, a <= x]
  --     right = quicksort [a | a <- xs, a > x]
  let left = quicksort $ filter (<= x) xs
      right = quicksort $ filter (> x) xs
  in
    left ++ [x] ++ right

-- take 10 $ quicksort [4000, 3999..1]
